File: src/dataset/javascript/forth_exercism-ThomasZumsteg/claude/ClaudeSonnet4_forth_v2.js
function n_args(n, func) {
  return function(stack) {
    const stackLen = stack.length;
    if (stackLen < n)
      throw new Error("Stack empty");
    const args = stack.splice(stackLen - n, n);
    const result = func(...args);
    if (result.length > 0) {
      stack.push(...result);
    }
  }
}

export default class Forth {
  constructor() {
    this.stack = [];
    this.env = {};
    this.builtin = {
      "+": n_args(2, (a, b) => [a+b]),
      "-": n_args(2, (a, b) => [a-b]),
      "*": n_args(2, (a, b) => [a*b]),
      "/": n_args(2, (a, b) => {
        if(b === 0)
          throw new Error("Division by zero");
        return [Math.floor(a/b)];
      }),
      "dup": n_args(1, a => [a, a]),
      "drop": n_args(1, () => []),
      "swap": n_args(2, (a, b) => [b, a]),
      "over": n_args(2, (a, b) => [a, b, a]),
    }
  }
  
  evaluate(line) {
    if (line.startsWith(': ') && line.endsWith(' ;')){
      const tokens = line.slice(2, -2).split(" ");
      const head = tokens[0].toLowerCase();
      if(!isNaN(head))
        throw new Error("Invalid definition");
      this.env[head] = tokens.slice(1);
      return;
    }

    const tokens = line.split(" ");
    let i = 0;
    while(i < tokens.length) {
      const token = tokens[i++].toLowerCase();
      const num = Number(token);
      if(!isNaN(num)) {
        this.stack.push(num);
      } else if(token in this.env) {
        tokens.splice(i, 0, ...this.env[token]);
      } else if(token in this.builtin) {
        this.builtin[token](this.stack);
      } else {
        throw new Error('Unknown command');
      }
    }
  }
}

File: src/dataset/python/forth_exercism-ThomasZumsteg/claude/ClaudeSonnet4_forth_v2.py
class StackUnderflowError(Exception):
    pass


keywords = {
    '+': (2, lambda a, b: (a + b,)),
    '-': (2, lambda a, b: (a - b,)),
    '*': (2, lambda a, b: (a * b,)),
    '/': (2, lambda a, b: (a // b,)),
    'swap': (2, lambda a, b: (b, a)),
    'over': (2, lambda a, b: (a, b, a)),
    'dup': (1, lambda x: (x, x)),
    'drop': (1, lambda x: tuple()),
    }


def evaluate(input_data):
    tokens = []
    for line in input_data:
        tokens.extend(line.split())
    
    stack = []
    env = {}
    i = 0
    
    while i < len(tokens):
        token = tokens[i].lower()
        i += 1
        
        if token == ':':
            keyword = tokens[i].lower()
            i += 1
            if keyword.isdigit():
                raise ValueError("Variables cannot be numbers: " + keyword)
            
            definition = []
            while tokens[i] != ';':
                definition.append(tokens[i])
                i += 1
            i += 1  # skip ';'
            env[keyword] = definition
            
        elif token in env:
            tokens[i:i] = env[token]
            
        elif token in keywords:
            n_args, func = keywords[token]
            if len(stack) < n_args:
                raise StackUnderflowError('Cannot apply ' + token)
            
            if n_args == 1:
                result = func(stack.pop())
            else:  # n_args == 2
                b = stack.pop()
                a = stack.pop()
                result = func(a, b)
            
            stack.extend(result)
            
        else:
            stack.append(int(token))
            
    return stack